File: algorithms/selection_sort/selection_sort.py
# find the smallest item

def findSmallest(arr):
    smallest_number = arr[0]
    smallest_number_index = 0
    
    for i in range(1,len(arr)):
        if arr[i] < smallest_number:
            smallest_number = arr[i]
            smallest_number_index = i
            
    return smallest_number_index

def selectionSort(arr):
    newArray = []
    
    for i in range(1,len(arr)):
        smallestElement = findSmallest(arr)
        
        # remove element from old array and add element to new array
        newArray.append(arr.pop(smallestElement))
    
    return newArray




print(selectionSort([4,3,6,1]))
        
    
    

File: algorithms/selection_sort/selection_sort.java
package algorithms.selection_sort;

public class selection_sort {

    public int[] selectionSort(int[] arr) {

        int i = 0;

        while (i < arr.length - 1) {

            int smallestNumberIndex = i;

            for (int j = i + 1; j < arr.length; i++) {

                if (arr[j] < arr[smallestNumberIndex]) {
                    smallestNumberIndex = j;
                }
            }

            int temp = arr[i];
            arr[i] = arr[smallestNumberIndex];
            arr[smallestNumberIndex] = temp;

            i++;

        }
        return arr;

    }

    public int[] anotherImplementation(int[] arr) {

        for (int i = 0; i < arr.length - 1; i++) {

            int smallestNumberIndex = i;

            for (int j = i + 1; j < arr.length; j++) {
                if (arr[j] < arr[smallestNumberIndex]) {
                    smallestNumberIndex = j;
                }
            }

            int temp = arr[i];
            arr[i] = arr[smallestNumberIndex];
            arr[smallestNumberIndex] = temp;
        }

        return arr;
    }

    // The outer loop iterates over each element of the array except the last one
    // (since the last element will be naturally sorted by the time all others are
    // sorted).

    // The inner loop finds the index of the smallest element in the unsorted part
    // of the array.

    // The smallest element found is then swapped with the first unsorted element.

    // The sorting is done in-place, meaning the original array arr is sorted
    // without the need for an additional array.

}
